### PR TITLE
[DPTP-4239] Verifying branch protection on public repos or paid plans

### PR DESCRIPTION
**Pull request by hector-vido**
This modifications verify if branch protection is enable for private repositories using a free plan, if this is the case, the verification should return an error.

### CODE REVIEW

File: cmd/check-gh-automation/main.go [hunk 1]
```diff
@@ -191,6 +200,11 @@ func checkRepos(repos []string, bots []string, appName string, ignore sets.Set[s
 			repoLogger.Info("no bots provided to check")
 		}
 
+		fullOrg, err := client.GetOrg(org)
+		if err != nil {
+			logger.Errorf("Error checking organization %s: %v", org, err)
+			return nil, fmt.Errorf("error checking organization %s: %w", org, err)
```
Comment (smg247) [inline cmd/check-gh-automation/main.go:206]:
```suggestion
			logger.Errorf("Error obtaining org from github: %s: %v", org, err)
			return nil, fmt.Errorf("error obtaining org from github: %s: %w", org, err)
```

File: cmd/check-gh-automation/main.go [hunk 2]
```diff
@@ -206,8 +220,20 @@ func checkRepos(repos []string, bots []string, appName string, ignore sets.Set[s
...
 
 			// If branch protection is configured, verify admin access for the hardcoded admin bot
+			// also branch protection should be used only on public repos or paid plans
 			if branchProtectionEnabled {
 				logger.Infof("Branch protection is enabled for %s/%s", org, repo)
+
+				if err != nil {
+					logger.Errorf("Error gathering repository data in %s/%s: %v", org, repo, err)
+					return nil, fmt.Errorf("error gathering repository data in %s/%s: %w", org, repo, err)
+				}
```
Comment (smg247) [inline cmd/check-gh-automation/main.go:230]: This error doesn't make sense...what is throwing an error here?

File: cmd/check-gh-automation/main.go [hunk 3]
```diff
@@ -206,8 +220,20 @@ func checkRepos(repos []string, bots []string, appName string, ignore sets.Set[s
...
+				if fullRepo.Private && fullOrg.Plan.Name == "free" {
```
Comment (smg247) [inline cmd/check-gh-automation/main.go:231]: Just thought of this, but is it possible for a repo to have its own "plan"? We should verify this prior to merging...

File: cmd/check-gh-automation/main.go [hunk 4]
```diff
@@ -206,8 +220,20 @@ func checkRepos(repos []string, bots []string, appName string, ignore sets.Set[s
...
+					failing.Insert(orgRepo)
+					logger.Errorf("Branch protection is enabled, the org %s has a free plan, the repository %s must be public", org, repo)
+					return nil, fmt.Errorf("branch protection is enabled, the org %s has a free plan, the repository %s must be public", org, repo)
```
Comment (smg247) [inline cmd/check-gh-automation/main.go:234]: We shouldn't just return here. Just log and add the repo to the `failing` set.

File: cmd/check-gh-automation/main_test.go [hunk 5]
```diff
@@ -329,6 +369,24 @@ func TestCheckRepos(t *testing.T) {
...
+			expected: []string{},
+		},
+		{
+			name:     "public repository has branch protection enabled and is a free plan",
+			repos:    []string{"org-5/repo-a"},
+			mode:     standard,
+			expected: []string{},
+		},
+		{
+			name:        "private repository has branch protection enabled and is not a free plan",
```
Comment (smg247) [inline cmd/check-gh-automation/main_test.go:385]: Is this test named correctly? Seems like it **is** the free plan. Either way, you should have tests for both cases.

File: cmd/check-gh-automation/main.go (hunk 2 above)
Comment (hector-vido) [reply cmd/check-gh-automation/main.go:230]: This is a left over of the other modifications I was doing, thanks for noticing!

File: cmd/check-gh-automation/main.go (hunk 3 above)
Comment (hector-vido) [reply cmd/check-gh-automation/main.go:231]: There is no "plan" property on "get a repository" api endpoint:
https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#get-a-repository

Also, the [Github documentation](https://docs.github.com/en/get-started/learning-about-github/githubs-plans) says that is only available for accounts:
```GitHub offers free and paid plans for storing and collaborating on code. Some plans are available only to personal accounts, while other plans are available only to organization and enterprise accounts.```

File: cmd/check-gh-automation/main_test.go (hunk 5 above)
Comment (hector-vido) [reply cmd/check-gh-automation/main_test.go:385]: No is not, thanks for noticing!

File: cmd/check-gh-automation/main.go (hunk 4 above)
Comment (hector-vido) [reply cmd/check-gh-automation/main.go:234]: True, this was given false positive on one of the tests!

File: cmd/check-gh-automation/main_test.go [hunk 6]
```diff
@@ -329,6 +372,24 @@ func TestCheckRepos(t *testing.T) {
...
+			mode:     standard,
+			expected: []string{},
+		},
+		{
+			name:     "private repository has branch protection enabled and is a free plan",
+			repos:    []string{"org-5/repo-d"},
+			mode:     standard,
+			expected: []string{"org-5/repo-d"},
+		},
 	}
```
Comment (smg247) [inline cmd/check-gh-automation/main_test.go:393]: Can we add a "private repository has branch protection enabled and is a pro plan" case too?

**Pull request merged**